File: cs/algorithms_in/binary_search.cs
namespace algorithms_in {
    public static class BinarySearch {
        public static int binarySearch(int[] array, int query) {
            int l = 0;
            int r = array.Length - 1;

            int i = l + (r - l) / 2;

            while (l <= r) {
                if (array[i] == query) return i;
                if (array[i] < query) l = i + 1;
                if (array[i] > query) r = i - 1;
                i = l + (r - l) / 2;
            }
            return -1;
        }
    }
}

File: rust/algorithms_in/src/binary_search.rs
pub fn binary_search(slice: &[usize], query: usize) -> i32 {
    let mut l: usize = 0;
    let mut r: usize = slice.len() - 1;

    let mut i: usize = l + (r - l) / 2;

    while l <= r {
        if slice[i] == query { return i.try_into().unwrap(); }
        if slice[i] < query { l = i + 1; }
        if slice[i] > query { r = i - 1; }
        if l == r { break; }
        i = l + (r - l) / 2;
    }
    return -1;
}